qutebrowser/browser/webengine/webenginetab.py

```diff
@@ -156,52 +156,83 @@ def to_printer(self, printer, callback=None):
+                                    findTextFinished signal.
+    """
+
+    def __init__(self):
+        self.active_match = 0
+        self.total_matches = 0
+        self.store_match_data_connected = False
+        self.flag_wrap = True
+
+    def store_match_data(self, result):
+        """Store information on the last match.
+
+        The information will be checked against when wrapping is turned off.
+
+        Args:
+            result: A FindTextResult passed by the findTextFinished signal.
+        """
+        self.active_match = result.activeMatch()
+        self.total_matches = result.numberOfMatches()
+        log.webview.debug("Active search match: {}/{}"
+                          .format(self.active_match, self.total_matches))
+
+    def reset_match_data(self):
+        """Reset match information.
+
+        Stale information could lead to next_result or prev_result misbehaving.
+        """
+        self.active_match = 0
+        self.total_matches = 0
+
+    def message_wrap_prevented(self, top=True):
+        """Show a message informing the user that wrapping was prevented."""
+        if top:
+            message.info("Search hit TOP")
+        else:
+            message.info("Search hit BOTTOM")
+
+
 class WebEngineSearch(browsertab.AbstractSearch):
 
     """QtWebEngine implementations related to searching on the page.
 
     Attributes:
         _flags: The QWebEnginePage.FindFlags of the last search.
-        _flag_wrap: An additional flag indicating whether the last search
-                    used wrapping.
         _pending_searches: How many searches have been started but not called
                            back yet.
-        _nowrap_available: Whether the nowrap functionality is available.
-        _active_match: The 1-based index of the currently active matc
-                       on the page.
-        _total_matches: The total number of search matches on the page.
     """
 
     def __init__(self, tab, parent=None):
         super().__init__(tab, parent)
         self._flags = QWebEnginePage.FindFlags(0)  # type: ignore
-        self._flag_wrap = True
         self._pending_searches = 0
         # The API necessary to stop wrapping was added in this version
         if qtutils.version_check("5.14"):
-            self._active_match = 0
-            self._total_matches = 0
-            self._nowrap_available = True
-            self._store_match_data_connected = False
+            self._wrap_handler = _WebEngineSearchWrapHandler()
         else:
-            self._nowrap_available = False
-
-    def _store_match_data(self, result):
-        self._active_match = result.activeMatch()
-        self._total_matches = result.numberOfMatches()
-        log.webview.debug("active match: number {0} of {1}"
-                          .format(self._active_match, self._total_matches))
-
-    def _reset_match_data(self):
-        self._active_match = 0
-        self._total_matches = 0
+            self._wrap_handler = None
```

Comment:
After some more thought (after writing the messages below), I'm wondering whether it would make more sense to have the wrap handler there whether it works or not (essentially the [null object pattern](https://en.wikipedia.org/wiki/Null_Object_pattern)).

Then the only thing that would change is whether the signal is connected or not, and the wrap handler can decide by itself whether it can act on the calls it is given, or simply ignores them.

That way, you could save a lot of `if self._wrap_handler is not None:` checks below.

What do you think?